ensime_shared/ensime.py

```diff
@@ -584,9 +584,13 @@ def handle_completion_info_list(self, call_id, payload):
         """Handler for a completion response."""
         completions = payload["completions"]
         self.log("handle_completion_info_list: in")
-        self.suggestions = [completion_to_suggest(c) for c in completions]
+        self.suggestions = [completion_to_suggest(c, self.format_type_params()) for c in completions]
         self.log("handle_completion_info_list: {}".format(self.suggestions))
 
+    def format_type_params(self):
+        """NOP as it's not available for Server 1.0"""
+        return lambda _: ""
+
```

Comment:
Actually this particular case of `format_type_params` might be a bad example, maybe it is best to duck-type it like you've done here so that code like `def completion_to_suggest(completion, formatted_type_params)` can be part of the common interface with the same signature. Or maybe it could be in kwargs to be more flexible to both? I'm not considering the specifics *too* closely just yet.

But overall, yeah, composition usually proves better than inheritance—roughly what I had in mind is a bit of both:

- Extract existing handler stuff from `EnsimeClient` to a mixin, called something like `ProtocolHandlerV1` maybe?
- Have `ProtocolHandlerV2` inherit from that, in Scala terms `trait ProtocolHandlerV2 extends ProtocolHandlerV1`. It can add stuff for new features and override for protocol changes as needed. I suppose there ought to be an abstract `ProtocolHandler` defining the interface, it's a bunch of OOP boilerplate but it's probably the proper thing to do even if Python can't statically check compliance 😄 
- Instantiate an `EnsimeClient` at startup dynamically mixing in the appropriate `ProtocolHandler` at runtime.

Then `EnsimeClient` could catch `NotImplementedError` in a common place, to log/inform the user that a new feature isn't supported in ENSIME v1.

This would introduce a good deal more OOP machinery, but on the other hand I think it would be a big step for better testability too. The worst thing deterring me from adding more tests currently is the tight coupling between `EnsimeClient`, a Vim instance (`self.vim = vim`), the launcher, and the handler logic which is the most important thing to cover with tests. If we can get to the point that those handler mixins can be tested in isolation, perhaps with just a minimal mock `vim` interface like @ChrisCoffey started in the test harness, that would be huge to me.

These are just my thoughts, some of the "architecture" stuff that's been bouncing around my head even before this server v2 discussion. Keen to hear everyone else's thoughts.